adlfs/core.py

```diff
@@ -743,6 +743,9 @@ def _put_data(rest, op, path, data, max_attempts=10, **kwargs):
         except (PermissionError, FileNotFoundError):
             raise
         except Exception as e:
+            if '"BadOffsetException"' in repr(e):
```

Comment:
@martindurant I think we are missing the purpose of including the offset parameter in the first place. It is strictly forbidden to append at any offset other than the very end of the file. You will get an exception if you try to append:
a) an index where there is already data
b) at an index past the end of the file.

The purpose of utilizing the offset parameter is to protect the multi-threaded upload portion from inadvertently uploading too much data without just failing out. Here is the specific scenario we are trying to protect against:
1. multi-part uploader begins upload 1GB file
2. while appending a 4mb block the front end doesn't respond back, so we cancel the connection and start the retry logic
3. What we don't know is that the append succeeded in the back end. The data is there and in the file.
4. If we append now (as a retry) without including the offset we expect the file to be at, then the append will succeed (again) and we will get an extra 4mb of data in the file.

The solution for this particular problem, is that applications that use the client can protect themselves from this scenario by explicitly specifying the offset they want, then making a decision about what to do if they get a badoffset exception.

The client itself should only give the caller options, it should not make any decisions for the caller directly.